### ISSUE TITLE
Key / config formatting

### ISSUE THREAD

**Issue by dcarosone** [oxy-secure/oxy#129]
I have just spend longer than I would care to admit failing to realise how to make server and client key/configs out of the `keygen` command. I started out thinking it would help if the `keygen` output was in two parts, with a separator and comments for which lines were supposed to go on server/client sections. Only when I started trying to mock it up, did I realise that the keys it generates can be used for either server or client.

I have no idea why I didn't realise that, but it seems like a UX trap that we could avoid. People will mess up private/public keys even in the best of circumstances.

Some thoughts, to factor into eventual rework of this stuff:
* distinguish server/client public/private keys, by config-field name, so there are 4 separate names and each can only go in the one place.
* have keygen take an arg for whether you want server or client keys, and output example configs for server and client files using that key
* consider having the internal format of the key values also encode (ideally visibly) their types, to make it harder to put them in the wrong places
* consider doing the above with explicit types and serialisations (basically, extend type safety all the way from config to consumption and usage)
* extend these last 2 points to the other keys (psk/knock/etc) that are already distinct by label and are shared.

There's also room for potential confusion in the fact that server files have `[[clients]]` stanzas and vice-versa.  Of course it makes sense once everything clicks, but before then it can contribute to accessibility difficulties and potentially dangerous mistakes. Using terms that are more intentional rather than declarative for some of the field names might help.

Putting all this together in an (overwrought for emphasis, highly strawman)  example:

server config:
```
[[this-server]]
authenticate-as-server-with = "OXY-SVR-PRV-GBJQEAIBGACQMAZLMVYAIIQEEA6VNDALVMYY46LAMUXO2C4K2JDFKJFZB5FNVGPCGGBUELFCVW7WVIJDAMQQBJVNDXQRHEC4W5CSBVALHXE4KZFZZIAE2KJC6Y5STBK55IFYO5TD"
knock = "OXY-KNOCK-MEPBVZ4332V23CJOOFTA"
psk = "OXY-PSK-6E6WGDJDDN5AU7JEQGZA"

[[allowed-client]]
name = "jmage"
client-identified-by = "OXY-USR-PUB-AU6MM3AQTMIPBBH26Q5YTIRDVIPIJHFKBKCKURJN4DVHH5XNBXIA"
setuser = "jmage"
```

```
[[connect-to-server]]
name = "banana"
server-identified-by = "OXY-SVR-PUB-U2WR3YITSBOLORJA2QFT3SOFMS44UACNFERPMOZJQVO6UC4HOZRQ"
authenticate-as-client-with = "OXY-USR-PRV-GBJQEAIBGACQMAZLMVYAIIQEECVJ562PXCBMQXIEBCKDUZGIDPXJRHN47BXANCJB64XH3NEHKOHRJIJDAMQQABJ4YZWBBGYQ6CCPV5B3RGRCHKQ6QSOKUCUEVJCS3YHKOP3O2DOQ"
psk = "OXY-PSK-6E6WGDJDDN5AU7JEQGZA" 
knock = "OXY-KNOCK-MEPBVZ4332V23CJOOFTA"
```

Finally, looking at `psk` and `knock` values, are they ever used separately (config-wise), or could they be combined to a single token?

**Comment by jennamagius** [comment]
Yeah, config files were added pretty late and I agree there's a ton to be done to make it easier to understand if you don't preexistingly understand.

PSK and knock values... knock values are server wide and PSKs are per-user when everything's set up the best way. So a server config would be like:

```
knock = "ABC"
[[clients]]
name="Samantha"
psk = "DEF"

[[clients]]
name="Jennifer"
psk="GHI"
```

So for that reason I think there's value in keeping them separate.

As to your dramatically more explicit config files: I dig it.

**Comment by jennamagius** [comment]
The fact that a client is denoted by "[[clients]]" is kinda super weird if you don't preexistingly know TOML, but that TOML document in my previous comment converts exactly to the following JSON:

{
  "knock": "ABC",
  "clients": [
    { "name": "Samantha", "psk": "DEF" },
    { "name": "Jennifer", "psk": "GHI" }
  ]
}

**Comment by jennamagius** [comment]
Just as a side note while we're talking about config files:

There are many, many cases where things that are currently sourced from arg:: need to be sourced from conf:: instead. The general pattern I've kinda built is that: conf will pass forward the value from arg, if present, and then take it from conf otherwise.

Really, basically any call to arg::matches() outside of conf.rs should be replaced with a call to conf.

**Comment by dcarosone** [comment]
Actually, for more symmetry and obvious association of field/value 

```
user-identified-by = "OXY-USR-ID-.."
server-identified-by = "OXY-SVR-ID-.."

authenticate-as-user-with: "OXY-USR-AUTH-..."
authenticate-as-server-with: "OXY-SVR-AUTH-..."
```

but I really just wanted to convey the general concept rather than bikeshed specific choices, especially with myself :roll_eyes:

**Comment by jennamagius** [comment]
The next thing I intend to do on this front is make an interactive configuration wizard that builds config files via a process like:

```
[jmage@host ~]$ oxy configure
Hello! There are currently no config files detected. Would you like to configure the sever or the client?
client
Okay, setting up the client. What's the name of the server you would like to configure your client to connect to?
banana
What's the public key of that server?
ASDF123
What's the knock key for that server?
ASDF123
Okay! That client profile is registered.
Your public key when connecting to this server will be: {}, 
Your knock value when connecting to this server will be: {},
```

The wizard will, of course, be completely optional and just manage a TOML document on the back end, but it makes it so: you can run the wizard a few times until you get a feel for what it's doing, then after you're a seasoned pro you won't need to use it anymore.

The other thing a wizard allows for conveniently is: #89 , which is a super important feature & the next thing I plan to work on.

**Comment by dcarosone** [comment]
I was thinking about the config file format overall, and in particular whether it might make sense to support something like a directory full of fragments, so that (for example) there can be a small server-config snippet file per user rather than edit one file.  They'd be semantically the same as additional sections in the main file. This is a common way to help automation and deployment scripts make idempotent and atomic changes without too much entanglement with the rest of the software.

A wizard that edits the config file could serve the same purpose, arguably even better if it deals with locking/atomicity and has a good interface to specify everything needed via args, and can take in snippets to validate and merge. I'd advocate that the tool (at least eventually) should be the preferred way of editing the config (a bit like useradd rather than editing /etc/passwd and 4 other things directly). This means the UI needs to be designed well enough to support that goal and encourage encrypted config use.

Another important thing to think about is who is running the tool for what purpose. A server admin mostly shouldn't be running the tool to generate keypairs for users; they should never see the privkey. Users should create themselves an id-blob, then that needs to be added to the server config, so in the current model with a single config file, they'd have to pass it to the server admin to add. That implies that the fragments being passed around should be validatable as a whole, maybe including a simple checksum for transport/copy-paste errors (this might sound a bit more like a pgp key blob, but that's not quite what I'm aiming for). Ideally, a user shouldn't have to ask the admin to add them, which means there's going to need to be a scheme with file fragments and ownership validation, or a gatekeeper admin command to do the editing.

I presume that encrypted config is also effectively a passphrase on user private keys, which would be good.

**Comment by dcarosone** [comment]
In the nearer term, I imagine you were thinking of something closer to the 'cargo edit' extension, which might be a great place to start, at least looking at some of the dependent crates it uses. The most recent release added some nice stuff like preserving toml formatting on changes.

**Comment by jennamagius** [comment]
Yes, encrypted config definitely does serve as a passphrase on private keys.

I definitely intend to make the wizard usable in "batch mode".

A ~/.config/server.conf.d/ type thing seems pretty reasonable to do.

Server admins should definitely never see user private keys. As to the paradigm where users remote-login enable themselves without talking to an admin... I think I don't like it. Like, that's the way SSH is,  where you expect there to be a master sshd process and any user can unilaterally opt-in to it by creating a ~/.ssh/authorized_keys. I'd rather an admin whitelist which users are able to leverage a given oxy server. Maybe the compromise there is like: server.conf can have sections like:

```
[[clients]]
setuser = "dcarsone"
include = "/home/dcarsone/.config/oxy/authorized_keys"
```

Or like a sugarier version:

```
# at the root of the config
user_confgs = [ "dcarsone" ]
```

Which would basically desugar to the more explicit one.

That'd give an approved user the ability to rotate their own keys without bugging an admin.

**Comment by jennamagius** [comment]
Yes, encrypted config definitely does serve as a passphrase on private keys.

I definitely intend to make the wizard usable in "batch mode".

A ~/.config/oxy/server.conf.d/ type thing seems pretty reasonable to do.

Server admins should definitely never see user private keys. As to the paradigm where users remote-login enable themselves without talking to an admin... I think I don't like it. Like, that's the way SSH is, where you expect there to be a master sshd process and any user can unilaterally opt-in to it by creating a ~/.ssh/authorized_keys. I'd rather an admin whitelist which users are able to leverage a given oxy server. Maybe the compromise there is like: server.conf can have sections like:

[[clients]]
setuser = "dcarosone"
include = "/home/dcarosone/.config/oxy/authorized_keys"

Or a sugarier version:

# at the root of the config
user_confgs = [ "dcarosone" ]

Which would basically desugar to the more explicit one.

That'd give an approved user the ability to rotate their own keys without bugging an admin.

**Comment by jennamagius** [comment]
I dunno, maybe it should just automatically pull in the user configs for every user. I've definitely established persistence inside a service account on pentests by making a ".ssh/authorized_keys" in /var (because that's where the service account's home directory was).

I'm undecided.

**Comment by BlackHoleFox** [comment]
I like the option of the "sugarier" version. It doesn't let just any user be a client but requires minimal administrative intervention incase they loose a key or whatnot.

**Comment by dcarosone** [comment]
It's complicated. It can't just be "pull in arbitrary config fragment", even if the scope is specific to the user based on some filename/path scheme. There are clearly admin-controlled restrictions (allowed commands, port-forwards, source ip's or whatever else we dream up) as well as things we might want the user to be able to do themselves. There will need to be some scheme of precedence where users can't configure less-restrictive settings than the admin.

That's a lot of complexity, will depend in part on the specifics of each option that will take time to shake out. Experience with ssh will give a lot of guidance but we don't just want to replicate that out of familiarity.

I suggest parking it and keeping the config admin-controlled for now, while keeping a mechanism for user preferences in mind.  In practice, users can always run their own server instance, and that's likely to be the most common deployment for a while yet.

**Comment by dcarosone** [comment]
> making a ".ssh/authorized_keys" in /var (because that's where the service account's home directory was)

And yeah, I'm not sure file permissions are a rich enough security model for all this, given that this gets past sshd's path permissions checks.

**Comment by jennamagius** [comment]
A great strength about trying to funnel people through the wizard is it gives us an opportunity to prevent users (or admins) from re-using keys for multiple purposes.

A common usage scenario for SSH is: a user has a single ~/.ssh/id_rsa + ~/.ssh/id_rsa.pub, and they disseminate that id_rsa.pub to many hosts. I extremely don't want that. Every different host should have a different long term client key.

The only practical way I see to funnel people away from that habit is by driving configuration through the wizard.

**Comment by dcarosone** [comment]
Decoupling private key passphrase protection from individual keys will help
here. That said, if a private key gets compromised, probably all of its
siblings on a given client probably will be too, if nothing else because of
passphrase reuse.

Oh, and whatever encrypting format gets used for config should of course
include integrity protection, which can cover the other config values as
well.

On 1 Jul. 2018 08:09, "jennamagius" <notifications@github.com> wrote:

A great strength about trying to funnel people through the wizard is it
gives us an opportunity to prevent users (or admins) from re-using keys for
multiple purposes.

A common usage scenario for SSH is: a user has a single ~/.ssh/id_rsa +
~/.ssh/id_rsa.pub, and they disseminate that id_rsa.pub to many hosts. I
extremely don't want that. Every different host should have a different
long term client key.

The only practical way I see to funnel people away from that habit is by
driving configuration through the wizard.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/oxy-secure/oxy/issues/129#issuecomment-401569215>, or mute
the thread
